// @opentelemetry/exporter-zipkin/build/src/zipkin.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZipkinExporter = void 0;
const api = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
const index_1 = require("./platform/index");
const transform_1 = require("./transform");
const resources_1 = require("@opentelemetry/resources");
/**
 * Zipkin Exporter
 */
class ZipkinExporter {
    constructor(config = {}) {
        this.DEFAULT_SERVICE_NAME = 'OpenTelemetry Service';
        this._sendingPromises = [];
        const urlStr = config.url || ZipkinExporter.DEFAULT_URL;
        this._logger = config.logger || new api.NoopLogger();
        this._send = index_1.prepareSend(this._logger, urlStr, config.headers);
        this._serviceName = config.serviceName;
        this._statusCodeTagName = config.statusCodeTagName || transform_1.statusCodeTagName;
        this._statusDescriptionTagName =
            config.statusDescriptionTagName || transform_1.statusDescriptionTagName;
        this._isShutdown = false;
    }
    /**
     * Export spans.
     */
    export(spans, resultCallback) {
        if (typeof this._serviceName !== 'string') {
            this._serviceName = String(spans[0].resource.attributes[resources_1.SERVICE_RESOURCE.NAME] ||
                this.DEFAULT_SERVICE_NAME);
        }
        this._logger.debug('Zipkin exporter export');
        if (this._isShutdown) {
            setTimeout(() => resultCallback({
                code: core_1.ExportResultCode.FAILED,
                error: new Error('Exporter has been shutdown'),
            }));
            return;
        }
        const promise = new Promise(resolve => {
            this._sendSpans(spans, this._serviceName, result => {
                resolve();
                resultCallback(result);
                const index = this._sendingPromises.indexOf(promise);
                this._sendingPromises.splice(index, 1);
            });
        });
        this._sendingPromises.push(promise);
    }
    /**
     * Shutdown exporter. Noop operation in this exporter.
     */
    shutdown() {
        this._logger.debug('Zipkin exporter shutdown');
        this._isShutdown = true;
        return new Promise((resolve, reject) => {
            Promise.all(this._sendingPromises).then(() => {
                resolve();
            }, reject);
        });
    }
    /**
     * Transform spans and sends to Zipkin service.
     */
    _sendSpans(spans, serviceName, done) {
        const zipkinSpans = spans.map(span => transform_1.toZipkinSpan(span, String(span.attributes[resources_1.SERVICE_RESOURCE.NAME] ||
            span.resource.attributes[resources_1.SERVICE_RESOURCE.NAME] ||
            serviceName), this._statusCodeTagName, this._statusDescriptionTagName));
        return this._send(zipkinSpans, (result) => {
            if (done) {
                return done(result);
            }
        });
    }
}
exports.ZipkinExporter = ZipkinExporter;
ZipkinExporter.DEFAULT_URL = 'http://localhost:9411/api/v2/spans';
