// @opentelemetry/exporter-zipkin/build/src/platform/browser/util.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareSend = void 0;
const core_1 = require("@opentelemetry/core");
/**
 * Prepares send function that will send spans to the remote Zipkin service.
 */
function prepareSend(logger, urlStr, headers) {
    let xhrHeaders;
    const useBeacon = navigator.sendBeacon && !headers;
    if (headers) {
        xhrHeaders = Object.assign({ Accept: 'application/json', 'Content-Type': 'application/json' }, headers);
    }
    /**
     * Send spans to the remote Zipkin service.
     */
    return function send(zipkinSpans, done) {
        if (zipkinSpans.length === 0) {
            logger.debug('Zipkin send with empty spans');
            return done({ code: core_1.ExportResultCode.SUCCESS });
        }
        const payload = JSON.stringify(zipkinSpans);
        if (useBeacon) {
            sendWithBeacon(payload, done, urlStr, logger);
        }
        else {
            sendWithXhr(payload, done, urlStr, logger, xhrHeaders);
        }
    };
}
exports.prepareSend = prepareSend;
/**
 * Sends data using beacon
 * @param data
 * @param done
 * @param urlStr
 * @param logger
 */
function sendWithBeacon(data, done, urlStr, logger) {
    if (navigator.sendBeacon(urlStr, data)) {
        logger.debug('sendBeacon - can send', data);
        done({ code: core_1.ExportResultCode.SUCCESS });
    }
    else {
        done({
            code: core_1.ExportResultCode.FAILED,
            error: new Error(`sendBeacon - cannot send ${data}`),
        });
    }
}
/**
 * Sends data using XMLHttpRequest
 * @param data
 * @param done
 * @param urlStr
 * @param logger
 * @param xhrHeaders
 */
function sendWithXhr(data, done, urlStr, logger, xhrHeaders = {}) {
    const xhr = new window.XMLHttpRequest();
    xhr.open('POST', urlStr);
    Object.entries(xhrHeaders).forEach(([k, v]) => {
        xhr.setRequestHeader(k, v);
    });
    xhr.onreadystatechange = () => {
        if (xhr.readyState === XMLHttpRequest.DONE) {
            const statusCode = xhr.status || 0;
            logger.debug('Zipkin response status code: %d, body: %s', statusCode, data);
            if (xhr.status >= 200 && xhr.status < 400) {
                return done({ code: core_1.ExportResultCode.SUCCESS });
            }
            else {
                return done({
                    code: core_1.ExportResultCode.FAILED,
                    error: new Error(`Got unexpected status code from zipkin: ${xhr.status}`),
                });
            }
        }
    };
    xhr.onerror = msg => {
        core_1.globalErrorHandler(new Error(`Zipkin request error: ${msg}`));
        return done({ code: core_1.ExportResultCode.FAILED });
    };
    // Issue request to remote service
    logger.debug('Zipkin request payload: %s', data);
    xhr.send(data);
}
